runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -158,11 +160,18 @@ public void write(IToken token, Object data) throws OverwriteException {
         }, true);
 
         // Cache the successful write
-        if (!runtime.getParameters().isCacheDisabled()) {
+        if (!runtime.getParameters().isCacheDisabled() && cacheOption == CacheOption.WRITE_THROUGH) {
```

Comment:
Interestingly, this is not technically a write through cache, since the write happens to the log unit first, then is added into the cache. But, close enough

Reply:
True! I'm actually not aware of any terminology that captures this ordering. I thought about WRITE_AFTER , but I'm not sure if that would be more confusing. Let me know if you have a better term.

Reply:
You could make it write-through by using `cache.asMap().compute` to perform the operation under the cache's hash-bin lock (~ per key). For example, assuming `doWrite` returns the `LogData` param this could be,

```java
boolean cachable = !runtime.getParameters().isCacheDisabled() 
    && (cacheOption == CacheOption.WRITE_THROUGH);
if (cachable) {
  readCache.asMap().compute(token.getTokenValue(), (key, oldValue) -> doWrite(token, data));
} else {
  doWrite(token, ld);
}
```

If you are okay with always invoking through the cache, then you can simplify it down to,
```java
boolean cachable = !runtime.getParameters().isCacheDisabled() 
    && (cacheOption == CacheOption.WRITE_THROUGH);
readCache.asMap().compute(token.getTokenValue(), (key, oldValue) -> {
    LogData newValue = doWrite(token, data);
    return cachable ? newValue : oldValue;
});
```
where `oldValue` may be `null` if absent.

The benefits of a computation is that no thread interleaving can stampede the entry with a stale value on the `put`. But that's likely already handled by Corfu, so just an fyi rather than recommendation.

Reply:
@ben-manes Thanks for the fyi

is this any different than just using a CacheWriter like:

```java
final LoadingCache<Long, ILogData> readCache = Caffeine.<Long, ILogData>newBuilder()
            .maximumSize(runtime.getParameters().getNumCacheEntries())
            .expireAfterAccess(runtime.getParameters().getCacheExpiryTime(), TimeUnit.SECONDS)
            .expireAfterWrite(runtime.getParameters().getCacheExpiryTime(), TimeUnit.SECONDS)
            .recordStats()
            .writer(new CacheWriter<Long, ILogData>() {
                @Override
                public void write(@Nonnull Long key, @Nonnull ILogData value) {
                    writeLU(value);
                }

                @Override
                public void delete(@Nonnull Long key,
                                   @Nullable ILogData value,
                                   @Nonnull RemovalCause cause) {

                }
            })
```

Reply:
You're right, that's likely the same. The writer is called under the lock, too.

The potential difference is that your `layoutHelper` method (aka `writeLU(value)`) is using parameters that the cache might not have (depending on your implementation details). The `token` is set on the `LogData` during the `layoutHelper` call, but the cache key is `token.getTokenValue()`. Most likely you could set it earlier or recreate the token, but I don't know. So you have a capturing lambda, whereas the `CacheWriter` is non-capturing by only knowing the direct inputs. If you need extra state from the environment then the `compute` is more flexible, but if not then it is identical.